core/storage/translation/gae_models.py

```diff
@@ -0,0 +1,137 @@
+
+"""Models for machine translation."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.platform import models
+import utils
+
+(base_models,) = models.Registry.import_models(
+    [models.NAMES.base_model])
+
+datastore_services = models.Registry.import_datastore_services()
+
+
+class MachineTranslatedTextModel(base_models.BaseModel):
+    """Model for storing machine translations"""
+
+    # The untranslated text.
+    origin_text = datastore_services.StringProperty(required=True)
+    # The language code of the untranslated text.
+    source_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the translation.
+    target_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The translation.
+    translated_text = datastore_services.TextProperty(
+        required=True, indexed=False)
+
+    @classmethod
+    def create(
+            cls, source_language_code, target_language_code, origin_text,
+            translated_text):
+        """Creates a new MachineTranslatedTextModel instance and returns its
+        ID.
+
+        Args:
+            source_language_code: str. The language code of the untranslated
+                text.
+            target_language_code: str. The language code of the translation.
+            origin_text: str. The untranslated text.
+            translated_text: str. The translation.
+
+        Returns:
+            MachineTranslatedTextModel. The newly created
+            MachineTranslatedTextModel instance.
+        """
+
+        entity_id = cls._generate_id(
+            source_language_code, target_language_code, origin_text)
+        translation_entity = cls(
+            id=entity_id,
+            source_language_code=source_language_code,
+            target_language_code=target_language_code,
+            origin_text=origin_text,
+            translated_text=translated_text)
+        translation_entity.put()
+        return entity_id
+
+    @classmethod
+    def _generate_id(
+            cls, source_language_code, target_language_code, origin_text):
+        """Generates key for the instance of MachineTranslatedTextModel
+        class in the required format with the arguments provided.
+
+        Args:
+            source_language_code: str. The language code of the untranslated
+                text.
+            target_language_code: str. The language code of the translation.
+            origin_text: str. The untranslated text.
+
+        Returns:
+            str. The generated ID for this entity of the form
+            '[source_language_code]:[target_language_code]:[hashed_text]'.
+        """
+        # max_length is 450, less than maximum key length (500) to allow room
+        # for language_codes.
+        hashed_text = utils.convert_to_hash(origin_text, 450)
```

Comment:
Does utils.convert_to_hash() ever return something that's more than 450 characters? If not then it might make sense to use a smaller max length that's more "in range".

Reply:
I think utils.convert_to_hash() doesn't return anything over 50 characters. I changed it accordingly.